src/services/BlockchainScanner.ts

```diff
@@ -3,66 +3,112 @@ import {getModelForClass} from '@typegoose/typegoose';
+  dexProtocolName!: DexProtocolName;
+  step: number;
   provider: StaticJsonRpcProvider | undefined;
+  chainId!: ChainsIds;
+
+  constructor(chainId: ChainsIds, dexProtocolName: DexProtocolName, settings: Settings) {
+    this.step = 0;
+    this.logger = logger;
+
+    if (!chainId || !dexProtocolName) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] chainId or dexProtocolName not set`);
+    }
+
+    this.chainId = chainId;
+    this.dexProtocolName = dexProtocolName;
+
+    if (!settings.blockchains[chainId].providerUrl.join('')) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] provider URL not set`);
+    }
+
+    const startBlock = settings.dexes[chainId]?.[dexProtocolName]?.startBlock;
+
+    if (!startBlock || Number(startBlock) <= 0) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] startBlock must be higher than 0`);
+    }
+
+    this.startBlock = startBlock!;
+    this.step = settings.dexes[chainId]?.[dexProtocolName]?.agentStep || 0;
+  }
 
   abstract apply(fromBlock: number, toBlock: number): Promise<boolean>;
 
   async run(): Promise<Result> {
     const fromBlock = await this.getMarker();
 
     if (!this.provider) {
-      this.logger.info('[BlockchainScanner] not active');
+      throw new Error(`[BlockchainScanner][${this.chainId}][${this.dexProtocolName}] getToBlock: no provider`);
+    }
+
+    const latestBlock = await this.provider.getBlockNumber();
+    const stepBlock = fromBlock + this.step;
+    const blocksToScan = latestBlock - fromBlock;
+
+    if (blocksToScan < this.step) {
+      this.logger.debug(
+        // eslint-disable-next-line max-len
+        `[BlockchainScanner][${this.chainId}][${this.dexProtocolName}] fromBlock: ${fromBlock}, latestBlock: ${latestBlock}, step: ${this.step}, stepBlock: ${stepBlock}`,
+      );
+      return {fromBlock, success: false, waitBlockRange: true};
+    }
+
+    const toBlock = await this.getToBlock(fromBlock, latestBlock);
+
+    if (!toBlock) {
       return {success: false, fromBlock};
     }
 
-    const toBlock = await this.getToBlock(fromBlock);
     let success = false;
 
     try {
       success = await this.apply(fromBlock, toBlock);
     } catch (e) {
-      this.logger.error(`[BlockchainScanner] ${e}`);
+      console.log(`[BlockchainScanner][${this.chainId}][${this.dexProtocolName}] ${e}`);
     }
 
     if (success) await this.updateMarker(toBlock);
     return {fromBlock, toBlock, success, synchronized: toBlock < fromBlock + this.step};
   }
 
-  private async getToBlock(fromBlock: number): Promise<number> {
-    if (!this.provider) throw new Error('[BlockchainScanner] getToBlock: no provider');
-
-    const latestBlock = await this.provider.getBlockNumber();
+  private async getToBlock(fromBlock: number, latestBlock: number): Promise<number | undefined> {
     const stepBlock = fromBlock + this.step;
+
+    // Math.trunc(blocksToScan / this.step) * this.step;
```

Comment:
why this is commented out? this is correct math

Reply:
I left a comment here https://github.com/umbrella-network/pegasus/pull/347#discussion_r1573359232
Like, it is returning a number higher than the step, and step should be the max. Or I missed something